spec/app/ics-030-middleware/README.md

```diff
@@ -0,0 +1,237 @@
+
+#### Handshake Callbacks
+
+```typescript
+function onChanOpenInit(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string) {
+    middlewareVersion, appVersion = splitMiddlewareVersion(version)
+    doCustomLogic()
+    app.OnChanOpenInit(
+        order,
+        connectionHops,
+        portIdentifier,
+        channelIdentifier,
+        counterpartyPortIdentifier,
+        counterpartyChannelIdentifier,
+        appVersion, // note we only pass app version here
+    )
+}
+
+function OnChanOpenTry(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string,
+  counterpartyVersion: string) {
+      cpMiddlewareVersion, cpAppVersion = splitMiddlewareVersion(counterpartyVersion)
+      middlewareVersion, appVersion = splitMiddlewareVersion(version)
+      if !isCompatible(cpMiddlewareVersion, middlewareVersion) {
+          return error
+      }
+      doCustomLogic()
+
+      // call the underlying applications OnChanOpenTry callback
+      app.OnChanOpenTry(
+          order,
+          connectionHops,
+          portIdentifier,
+          channelIdentifier,
+          counterpartyPortIdentifier,
+          counterpartyChannelIdentifier,
+          cpAppVersion, // note we only pass counterparty app version here
+          appVersion, // only pass app version
+      )
+}
+
+function onChanOpenAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  version: string) {
+      middlewareVersion, appVersion = splitMiddlewareVersion(version)
+      if !isCompatible(middlewareVersion) {
+          return error
+      }
+      doCustomLogic()
+      
+      // call the underlying applications OnChanOpenTry callback
+      app.OnChanOpenAck(portIdentifier, channelIdentifier, appVersion)
+}
+
+function OnChanOpenConfirm(
+    portIdentifier: Identifier,
+    channelIdentifier: Identifier) {
+    doCustomLogic()
+
+    app.OnChanOpenConfirm(portIdentifier, channelIdentifier)
+}
+
+function splitMiddlewareVersion(version: string): []string {
+    splitVersions = split(version,  ":")
+    middlewareVersion = version[0]
+    appVersion = join(version[1:], ":")
+    return []string{middlewareVersion, appVersion}
+}
+```
+
+NOTE: Middleware that does not need to negotiate with a counterparty middleware on the remote stack will not implement the version splitting and negotiation, and will simply perform its own custom logic on the callbacks without relying on the counterparty behaving similarly.
+
+#### Packet Callbacks
+
+```typescript
+function onRecvPacket(packet: Packet, relayer: string): bytes {
+    doCustomLogic()
+
+    app_acknowledgement = app.onRecvPacket(packet, relayer)
```

Comment:
Where `app` comes from?I guess that what we mean here by app is actually anything implementing ICS26 interface (`ModuleCallbacks`) as we have potentially several wrappers chained together. If I am not wrong, we might need to make some changes to a router module to be able to support chain of wrappers.

Reply:
The app here is the underlying application that implements ICS26. It is embedded by the given middleware (how it is embedded is up to specific languages/implementation. In ibc-go, it would be a field in the middleware struct)

The router module should have no changes. The goal is for ICS-4 and the underlying app to be completely oblivious to the existence of the middleware. This is accomplished on the ICS-4 side, since it is still only talking to some ICS26 app which is bound to a unique port. In the middleware case, it is talking to some top-level middleware but it doesn't matter. It's just an ICS-26 app. The middleware itself will route the call further down the chain.